The `iss` value isn't validated when encoding payload

I was looking into what PyGitHub does and writing test for it, invoking its JWT methods. Then I attempted decoding that JWT with `jwt.decode(..., issuer=123)` since there was `'iss': 123` in the object. The validator treats anything non-`str` as iterables and crashes:

```python-traceback
    def _validate_iss(self, payload: dict[str, Any], issuer: Any) -> None:
        if issuer is None:
            return
    
        if "iss" not in payload:
            raise MissingRequiredClaimError("iss")
    
        if isinstance(issuer, str):
            if payload["iss"] != issuer:
                raise InvalidIssuerError("Invalid issuer")
        else:
>           if payload["iss"] not in issuer:
E           TypeError: argument of type 'int' is not iterable

issuer     = 123
payload    = {'exp': 1739881471, 'iat': 1739881111, 'iss': 123}
self       = <jwt.api_jwt.PyJWT object at 0x7f3101bc93a0>

```

When I attempted `jwt.decode(..., issuer='123')` it hit another code branch and crashed in the equality check:
```python-traceback
    def _validate_iss(self, payload: dict[str, Any], issuer: Any) -> None:
        if issuer is None:
            return
    
        if "iss" not in payload:
            raise MissingRequiredClaimError("iss")
    
        if isinstance(issuer, str):
            if payload["iss"] != issuer:
>               raise InvalidIssuerError("Invalid issuer")
E               jwt.exceptions.InvalidIssuerError: Invalid issuer

issuer     = '123'
payload    = {'exp': 1739881584, 'iat': 1739881224, 'iss': 123}
self       = <jwt.api_jwt.PyJWT object at 0x7ff5d86cd2e0>
```

I understand why this is happening and I'm not the one creating the payload with `iss` being of `int` type. I believe that the root cause is that PyGitHub creates it with an integer and the bug is that PyJWT allows it, not validating the input.

I checked the spec @ https://datatracker.ietf.org/doc/html/rfc7519#section-4.1 and it says this is supposed to be a string:
> [4.1.1](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.1).  "iss" (Issuer) Claim
>
>   The "iss" (issuer) claim identifies the principal that issued the
>   JWT.  The processing of this claim is generally application specific.
>   The "iss" value is **a case-sensitive string** containing a StringOrURI
>   value.  Use of this claim is OPTIONAL.

*(emphasis mine)*

So here we are — PyJWT allows putting arbitrary values into payload on encoding but expects them to be spec-compliant on decoding.

I think, there are two things necessary to maintain consistency here — input validation with `jwt.encode()` checking that the values provided are of legal types, and maybe better type checking and error messages in `jwt.decode()` reporting illegal data types.

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -11,7 +11,7 @@ Fixed
 ~~~~~
 - Validate key against allowed types for Algorithm family in `#964 <https://github.com/jpadilla/pyjwt/pull/964>`__
 - Add iterator for JWKSet in `#1041 <https://github.com/jpadilla/pyjwt/pull/1041>`__
-
+- Validate `iss` claim is a string during encoding and decoding by @pachewise in `#1040 <https://github.com/jpadilla/pyjwt/pull/1040>`__
 
 Added
 ~~~~~

--- a/jwt/api_jwt.py
+++ b/jwt/api_jwt.py
@@ -433,7 +433,9 @@ class PyJWT:
                 if iss not in issuer:
                     raise InvalidIssuerError("Invalid issuer")
             except TypeError:
-                raise InvalidIssuerError("Issuer param must be a str or list of str")
+                raise InvalidIssuerError(
+                    "Issuer param must be a str or list of str"
+                ) from None
 
 
 _jwt_global_obj = PyJWT()

--- a/jwt/api_jwt.py
+++ b/jwt/api_jwt.py
@@ -5,7 +5,7 @@ import warnings
 from calendar import timegm
 from collections.abc import Iterable, Sequence
 from datetime import datetime, timedelta, timezone
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING, Any, Container
 
 from . import api_jws
 from .exceptions import (
@@ -417,7 +417,7 @@ class PyJWT:
             raise InvalidAudienceError("Audience doesn't match")
 
     def _validate_iss(
-        self, payload: dict[str, Any], issuer: list[str] | str | None
+        self, payload: dict[str, Any], issuer: Container[str] | str | None
     ) -> None:
         if issuer is None:
             return
@@ -438,7 +438,7 @@ class PyJWT:
                     raise InvalidIssuerError("Invalid issuer")
             except TypeError:
                 raise InvalidIssuerError(
-                    "Issuer param must be a str or list of str"
+                    'Issuer param must be "str" or "Container[str]"'
                 ) from None
 
 

--- a/jwt/api_jwt.py
+++ b/jwt/api_jwt.py
@@ -412,19 +412,28 @@ class PyJWT:
         if all(aud not in audience_claims for aud in audience):
             raise InvalidAudienceError("Audience doesn't match")
 
-    def _validate_iss(self, payload: dict[str, Any], issuer: Any) -> None:
+    def _validate_iss(
+        self, payload: dict[str, Any], issuer: list[str] | str | None
+    ) -> None:
         if issuer is None:
             return
 
         if "iss" not in payload:
             raise MissingRequiredClaimError("iss")
 
+        iss = payload["iss"]
+        if not isinstance(iss, str):
+            raise InvalidIssuerError("Payload Issuer (iss) must be a string")
+
         if isinstance(issuer, str):
-            if payload["iss"] != issuer:
+            if iss != issuer:
                 raise InvalidIssuerError("Invalid issuer")
         else:
-            if payload["iss"] not in issuer:
-                raise InvalidIssuerError("Invalid issuer")
+            try:
+                if iss not in issuer:
+                    raise InvalidIssuerError("Invalid issuer")
+            except TypeError:
+                raise InvalidIssuerError("Issuer param must be a str or list of str")
 
 
 _jwt_global_obj = PyJWT()

--- a/jwt/api_jwt.py
+++ b/jwt/api_jwt.py
@@ -69,6 +69,10 @@ class PyJWT:
             if isinstance(payload.get(time_claim), datetime):
                 payload[time_claim] = timegm(payload[time_claim].utctimetuple())
 
+        # Issue #1039, iss being set to non-string
+        if "iss" in payload and not isinstance(payload["iss"], str):
+            raise TypeError("Issuer (iss) must be a string.")
+
         json_payload = self._encode_payload(
             payload,
             headers=headers,

--- a/tests/test_api_jwt.py
+++ b/tests/test_api_jwt.py
@@ -954,6 +954,20 @@ class TestJWT:
         with pytest.raises(InvalidJTIError):
             jwt.decode(token, secret, algorithms=["HS256"])
 
+    def test_validate_iss_with_container_of_str(self, jwt: PyJWT) -> None:
+        """Check _validate_iss works with Container[str]."""
+        payload = {
+            "iss": "urn:expected",
+        }
+        # pytest.mark.parametrize triggers Untyped Decorator mypy issue,
+        # so trying inline for now
+        for issuer in (
+            ["urn:expected", "urn:other"],
+            ("urn:expected", "urn:other"),
+            {"urn:expected", "urn:other"},
+        ):
+            jwt._validate_iss(payload, issuer=issuer)
+
     def test_validate_iss_with_non_str(self, jwt):
         """Regression test for #1039"""
         payload = {

--- a/tests/test_api_jwt.py
+++ b/tests/test_api_jwt.py
@@ -170,6 +170,16 @@ class TestJWT:
                 lambda t=t: jwt.encode(t, "secret", algorithms=["HS256"]),
             )
 
+    def test_encode_with_non_str_iss(self, jwt):
+        """Regression test for Issue #1039."""
+        with pytest.raises(TypeError):
+            jwt.encode(
+                {
+                    "iss": 123,
+                },
+                key="secret",
+            )
+
     def test_encode_with_typ(self, jwt):
         payload = {
             "iss": "https://scim.example.com",

--- a/tests/test_api_jwt.py
+++ b/tests/test_api_jwt.py
@@ -943,3 +943,19 @@ class TestJWT:
 
         with pytest.raises(InvalidJTIError):
             jwt.decode(token, secret, algorithms=["HS256"])
+
+    def test_validate_iss_with_non_str(self, jwt):
+        """Regression test for #1039"""
+        payload = {
+            "iss": 123,
+        }
+        with pytest.raises(InvalidIssuerError):
+            jwt._validate_iss(payload, issuer="123")
+
+    def test_validate_iss_with_non_str_issuer(self, jwt):
+        """Regression test for #1039"""
+        payload = {
+            "iss": "123",
+        }
+        with pytest.raises(InvalidIssuerError):
+            jwt._validate_iss(payload, issuer=123)

